src/dftbp/math/wignerseitz.F90

```diff
@@ -0,0 +1,216 @@
+  use dftbp_io_message, only : error
+  use dftbp_math_sorting, only : index_heap_sort
+  use dftbp_dftb_periodic, only : frac2cart
+
+  implicit none
+  private
+
+
+  public :: generateWignerSeitzGrid
+
+
+contains
+
+
+  !> Determines the lattice vectors pointing to all primitive unit cells that are either located
+  !! inside or on the surface of the Wigner-Seitz cell of the Born-von Karman supercell, defined by
+  !! an unshifted Monkhorst-Pack-type k-point grid, that is, the diagonal coefficients nKpt(i) of
+  !! the supercell folding matrix.
+  !!
+  !! The development of this routine was inspired by the pyscf code (under the Apache-2.0 License):
+  !! Sun, Q., Berkelbach, T.C., Blunt, N.S., Booth, G.H., Guo, S., Li, Z., Liu, J., McClain, J.D.,
+  !! Sayfutyarova, E.R., Sharma, S., Wouters, S. and Chan, G.K.-L. (2018),
+  !! PySCF: the Python-based simulations of chemistry framework.
+  !! WIREs Comput Mol Sci, 8: e1340.
+  !! DOI: 10.1002/wcms.1340
+  subroutine generateWignerSeitzGrid(nKpt, latVecs, wsVectors, wsSearchSize, distTol)
+
+    !> Supercell folding coefficients (diagonal elements, i.e. MP-type grid)
+    integer, intent(in) :: nKpt(:)
+
+    !> Real-space lattice vectors of periodic geometry
+    real(dp), intent(in) :: latVecs(:,:)
+
+    !> Lattice vectors inside/on the surface of the WS-supercell in fractional coordinates
+    integer, intent(out), allocatable :: wsVectors(:,:)
+
+    !> Optional index range to search for primitive unit cells inside the WS-supercell
+    integer, intent(in), optional :: wsSearchSize(:)
+
+    !> Optional tolerance for comparing real-space distances
+    real(dp), intent(in), optional :: distTol
+
+    !> Actual tolerance for comparing real-space distances
+    real(dp) :: distTol_
+
+    !! Actual index range to search for primitive unit cells inside the WS-supercell
+    integer :: wsSearchSize_(3)
+
+    !> Lattice vectors inside/on the surface of the WS-supercell in Cartesian coordinates
+    real(dp), allocatable :: wsVectorsCart(:,:)
+
+    !! Work buffer of real-space lengths in Cartesian coordinates
+    real(dp), allocatable :: wsLengths(:)
+
+    !! Index array for sorting operations
+    integer, allocatable :: ind(:)
+
+    !! Temporary distance index that indicates the central cell
+    integer :: iCenter
+
+    !! Auxiliary variables
+    integer :: n1, n2, n3, i1, i2, i3, ii, jj, diff(3), tmp3Int(3)
+    real(dp) :: dotProducts(3, 3), diffMat(3, 3)
+
+    !! Temporary storage for distances
+    real(dp), allocatable :: dist(:)
+
+    @:ASSERT(all(nKpt > 0))
+    @:ASSERT(size(latVecs, dim=2) == 3)
+
+    if (present(wsSearchSize)) then
+      @:ASSERT(size(wsSearchSize, dim=1) == 3)
+      wsSearchSize_(:) = wsSearchSize
+    else
+      wsSearchSize_(:) = 2
+    end if
+
+    if (present(distTol)) then
+      @:ASSERT(distTol > 0.0_dp)
+      distTol_ = distTol
+    else
+      distTol_ = 1.0e-08_dp
+    end if
+
+    ! Cache repeatedly used dot-products
```

Comment:
Isn't that just a GEMM (or even a SYMM) product between latVecs and latVects.T ?

Reply:
I might have missed that, let me think about it.

Reply:
So `dotProducts == transpose(dotProducts)` is true, but also `dotProducts /= matmul(latVecs, transpose(latVecs))`, if that is what you meant.